src/main/scala/scalismo/io/StatisticalModelIO.scala

```diff
@@ -342,4 +387,231 @@ object StatismoIO {
+      // We recover U from it.
+
+      val lambdaSqrt = pcaVarianceVector.map(l => math.sqrt(l).toFloat)
+      val lambdaSqrtInv = lambdaSqrt.map(l => if (l > 1e-8) 1.0f / l else 0f)
+
+      // The following code is an efficient way to compute: pcaBasisMatrix * breeze.linalg.diag(lambdaSqrtInv)
+      // (diag returns densematrix, so the direct computation would be very slow)
+      val U = DenseMatrix.zeros[Float](pcaBasisMatrix.rows, pcaBasisMatrix.cols)
+      for (i <- 0 until pcaBasisMatrix.cols) {
+        U(::, i) := pcaBasisMatrix(::, i) * lambdaSqrtInv(i)
+      }
+      U
+    }
+
+    val modelOrFailure = for {
+      h5file <- HDF5Utils.openFileForReading(file)
+
+      representerName <- h5file.readStringAttribute(s"$modelPath/representer/", "name")
+      // read mesh according to type given in representer
+      image <- representerName match {
+        case "itkStandardImageRepresenter" => readImageRepresenter(h5file, modelPath)
+        case _ =>
+          h5file.readStringAttribute(s"$modelPath/representer/", "datasetType") match {
+            case Success("IMAGE") => readImageRepresenter(h5file, modelPath)
+            case Success(datasetType) => Failure(new Exception(s"can only read model of datasetType IMAGE. Got $datasetType instead"))
+            case Failure(t) => Failure(t)
+          }
+      }
+
+      meanArray <- h5file.readNDArray[Float](s"$modelPath/model/mean")
+      meanVector = DenseVector(meanArray.data)
+      pcaBasisArray <- h5file.readNDArray[Float](s"$modelPath/model/pcaBasis")
+      majorVersion <- if (h5file.exists("/version/majorVersion")) h5file.readInt("/version/majorVersion")
+      else {
+        if (representerName == "vtkPolyDataRepresenter" || representerName == "itkMeshRepresenter") Success(0)
+        else Failure(new Throwable(s"no entry /version/majorVersion provided in statismo file."))
+      }
+      minorVersion <- if (h5file.exists("/version/minorVersion")) h5file.readInt("/version/minorVersion")
+      else {
+        if (representerName == "vtkPolyDataRepresenter" || representerName == "itkMeshRepresenter") Success(8)
+        else Failure(new Throwable(s"no entry /version/minorVersion provided in statismo file."))
+      }
+      pcaVarianceArray <- h5file.readNDArray[Float](s"$modelPath/model/pcaVariance")
+      pcaVarianceVector = DenseVector(pcaVarianceArray.data)
+      pcaBasisMatrix = ndArrayFloatToMatrix(pcaBasisArray)
+      pcaBasis <- (majorVersion, minorVersion) match {
+        case (1, _) => Success(pcaBasisMatrix)
+        case (0, 9) => Success(pcaBasisMatrix)
+        case (0, 8) => Success(extractOrthonormalPCABasisMatrix(pcaBasisMatrix, pcaVarianceVector)) // an old statismo version
+        case v => Failure(new Throwable(s"Unsupported version ${v._1}.${v._2}"))
+      }
+
+      _ <- Try {
+        h5file.close()
+      }
+    } yield {
+      // statismo stores the mean as the point position, not as a displacement on the reference.
```

Comment:
This comment seems to be a left over. Can't this be removed? Or should it be accounted for that in the following lines?